test/evp_extra_test.c

```diff
@@ -4562,6 +4562,11 @@ int setup_tests(void)
 # ifndef OPENSSL_NO_DYNAMIC_ENGINE
     /* Tests only support the default libctx */
     if (testctx == NULL) {
+#  ifdef _WIN32
+        _putenv("OPENSSL_engine_libcrypto_override=yes");
+#  else
+        setenv("OPENSSL_engine_libcrypto_override", "yes", 1);
+#  endif
```

Comment:
It's a pretty bad sign that this does seem to work with a static libcrypto in master and 3.0,
while 1.1.1 has the same test case, but a dynamic licrypto instead.
Basically we have a memory leak here, since nobody does fully initialize everything
to make sure that the atexit handler is actually called.

Reply:
It is actually a design flaw that OPENSSL_cleanup in the dynamic libcrypto is not called.
in 1.1.1 we have this, which ensures that OPENSSL_cleanup is finally called:
https://github.com/openssl/openssl/blob/4c5c2a5efbc315d7926cafbd5a19044ee3e087fa/crypto/err/err.c#L295-L310

in master we have this instead:
https://github.com/openssl/openssl/blob/57645339ab645fe5abffe14fc005b5402ce03b84/crypto/err/err.c#L206-L221

BUT OPENSSL_INIT_BASE_ONLY does not install the atexit handler, thus the allocated globals
are not always cleaned up, because that is done by OPENSSL_cleanup.

This was broken by 12b4e582

Reply:
> It's a pretty bad sign that this does seem to work with a static libcrypto in master and 3.0, while 1.1.1 has the same test case, but a dynamic licrypto instead. Basically we have a memory leak here, since nobody does fully initialize everything to make sure that the atexit handler is actually called.

Linux build of master has `-Wl,-znodelete` linker flag which prevent library unloading, so `atexit` cannot ever happen, even if it is installed.

Reply:
There is at least one other O/S, but I don't recall the name...

Reply:
> There is at least one other O/S, but I don't recall the name...

NonStop would need:
```
#    elif defined __TANDEM
         putenv("OPENSSL_engine_libcrypto_override=yes");
#    endif
````

Reply:
> Linux build of master has -Wl,-znodelete linker flag which prevent library unloading, so atexit cannot ever happen, even if it is installed.

FYI: there is also `./config no-pinshared`